src/compiler/checker.ts

```diff
@@ -7555,6 +7559,60 @@ namespace ts {
             return undefined;
         }
 
+        function shouldAcquireLiteralType(literalNode: StringLiteral) {
+
+            let current: Node = literalNode;
+            while (true) {
+                const { parent } = current;
+                switch (parent.kind) {
+                    // The operand of a 'switch' should get a literal type.
+                    case SyntaxKind.SwitchStatement:
+                        return current === (parent as SwitchStatement).expression;
+
+                    // The tested expression of a 'case' clause should get a literal type.
+                    case SyntaxKind.CaseClause:
+                        return current === (parent as CaseClause).expression;
+
+                    case SyntaxKind.BinaryExpression:
+                        const binaryExpr = parent as BinaryExpression;
+                        switch (binaryExpr.operatorToken.kind) {
+                            // Either operand of an equality/inequality comparison
+                            // should get a literal type.
+                            case SyntaxKind.EqualsEqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsEqualsToken:
+                            case SyntaxKind.EqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsToken:
+                                return current === binaryExpr.left || current === binaryExpr.right;
+
+                            case SyntaxKind.BarBarToken:
+                                current = parent;
+                                continue;
+
+                        }
+                        break;
+
+                    case SyntaxKind.ConditionalExpression:
+                    case SyntaxKind.ParenthesizedExpression:
+                        current = parent;
+                        continue;
+                }
+
+                // This is not a node we can account for.
+                // Let contextual typing take over.
+                break;
+            }
+
+            // We could perform our walk, check if 'current' is an expression
+            // when we get to an unsupported node, and then get the contextual type on that.
+            // That would save a few steps through "pass-though" nodes,
+            // but the checks in 'isExpression' seem so involved that it would probably
+            // be better to simply grab the contextual type first and then see if we
+            // need to do any work.
+            const contextualType = getContextualType(literalNode);
+            return !!contextualType && contextualTypeIsStringLiteralType(contextualType);

```

Comment:
Oh I get it now, thanks. My question was, why does the contextual type matter at all? And now I realize you do still care whether the contextual type specifies a string literal, for basic cases like this:
```ts
var s: "hello" = "hello";
```